src/geometry/types.rs

```diff
@@ -121,6 +121,73 @@ impl Vector {
     }
 }
 
+// Rectangle /////////////////////////////////////////////////////////////////
+
+pub struct Size {
+    pub width: f32,
+    pub heigh: f32,
+}
+
+pub struct Rect {
+    // the leftmost coordination system of the rectangle
+    pub origin: Point,
+    // the width and high of the rectangle
+    pub size: Size,
+}
+
+impl Rect {
+
+    pub fn half_size(&mut self) {
+        self.size.width = self.size.width/2.0;
+        self.size.heigh = self.size.heigh/2.0;
+    }
```

Comment:
Here's an idea to make the code more concise. We could replace this method `half_size()` with a private instance method that returns a new `Rect` with the size already smaller. 

For example: `fn quarter_sized(&self) -> Rect`

The origin would still be the same, but the size is already smaller. Then in the `quadrants` method you wouldn't need to use `clone()` anymore (you could even delete the implementation of `clone()`), you could just write:

`let mut top_left = self.quarter_sized();`

Reply:
i just thought about this idea, the size of the rectangle can be smaller, but if we use this method, it is just true for  a quarter of the rectangle with the coordinator , because the other 3 Rectangles have different coordinators

Reply:
ah we still can change the coordinator extra ^^